src/bin/glarea/glium_backend.rs

```diff
@@ -0,0 +1,170 @@
+//! # glium_backend utility module
+//!
+//! This module provides some general utility typoes that allow glium to manage
+//! the rendering of a GLArea.
+
+use std::cell::RefCell;
+use std::rc::Rc;
+
+use gtk::{GLAreaExt, WidgetExt};
+use shared_library::dynamic_library::DynamicLibrary;
+
+/// The primary trait of this module, implemented by users to facilitate
+/// glium rendering for their GLArea
+pub trait GliumRenderer {
+    /// Use this method to set up your GL resources. This is called during the
+    /// `realize` signal callback for the GLArea. At this point, a valid
+    /// OpenGL context for the GLArea has been set up and is available via
+    /// `Facade`.
+    fn initialize(&mut self, facade: &Facade, gl_area: &gtk::GLArea);
+
+    /// Use this method to tear down your GL resources. This is called during
+    /// the `unrealize` signal callback for the GLArea.
+    fn tear_down(&mut self, gl_area: &gtk::GLArea);
+
+    /// Use this method to draw to the given Frame. This is called during the
+    /// `render` signal callback for the GLArea.
+    fn draw(
+        &mut self,
+        frame: glium::Frame,
+        gl_area: &gtk::GLArea,
+        gl_context: &gdk::GLContext,
+    ) -> gtk::Inhibit;
+}
+
+/// Utility function showing how to use a GlAreaGliumHook to connect a GLArea to
+/// glium via a GliumRenderer.
+pub fn hook_to_renderer<R>(renderer: R, gl_area: &gtk::GLArea, check_current_context: bool)
+where
+    R: GliumRenderer + 'static,
+{
+    let hook = GlAreaGliumHook::new(renderer, check_current_context);
+
+    gl_area.connect_realize(glib::clone!(@strong hook => move |gl_area| {
+        hook.borrow_mut().realize(gl_area);
+    }));
+
+    gl_area.connect_unrealize(glib::clone!(@strong hook => move |gl_area| {
+        hook.borrow_mut().unrealize(gl_area);
+    }));
+
+    gl_area.connect_render(glib::clone!(@strong hook => move |gl_area, gl_context| {
+        hook.borrow_mut().render(gl_area, gl_context)
+    }));
+}
+
+/// This struct is used to hook a GLArea widget to glium using a GliumRenderer
+pub struct GlAreaGliumHook<R>
+where
+    R: GliumRenderer,
+{
+    renderer: R,
+    facade: Option<Facade>,
+    check_current_context: bool,
+}
+
+impl<R> GlAreaGliumHook<R>
+where
+    R: GliumRenderer,
+{
+    pub fn new(renderer: R, check_current_context: bool) -> Rc<RefCell<Self>> {
+        Rc::new(RefCell::new(Self {
+            renderer,
+            facade: None,
+            check_current_context,
+        }))
+    }
+
+    pub fn realize(&mut self, gl_area: &gtk::GLArea) {
+        let context = unsafe {
+            glium::backend::Context::new::<_>(
+                Backend::new(gl_area.clone()),
+                self.check_current_context,
+                Default::default(),
+            )
+            .unwrap() // TODO: Better error handling`
```

Comment:
TODO :)

Reply:
Replaced just with stderr logging, since the only more robust error handling mechanisms I could think of would distract from the example.